ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -215,11 +217,15 @@ addBlock cdb@CDB{..} b = do
       -> m ()
     scheduleChainSelection curSlot slot hdr = do
       nbScheduled <- atomically $ updateTVar cdbFutureBlocks $ \futureBlocks ->
-        let futureBlocks' = Map.insertWith (<>) slot (hdr NE.:| []) futureBlocks
+        let futureBlocks' = Map.insertWith strictAppend slot
+              (forceElemsToWHNF (hdr NE.:| [])) futureBlocks
             nbScheduled   = fromIntegral $ sum $ length <$> Map.elems futureBlocks
         in (futureBlocks', nbScheduled)
       trace $ ScheduledChainSelection (headerPoint hdr) curSlot nbScheduled
 
+    strictAppend :: (Semigroup (t a), Foldable t) => t a -> t a -> t a
+    strictAppend x y = forceElemsToWHNF (x <> y)
```

Comment:
This full traversal seems excessive. The elements have already been forced at this point, and in particular so has the spine of the old list.

 Maybe that's inconsequential in this context.

Reply:
This list will be quite short and bounded, it will be at most the number of chain sync clients * rate limit, and in practice mostly empty or <= the number of chain sync clients.

Maybe `seq`in on the `length` suffices, I'll try that.

Reply:
> Maybe `seq`in on the `length` suffices, I'll try that.

Doesn't work, `NoUnexpectedThunks` finds a `SelectorClosure`.

BTW, I was making the change not to avoid a space leak, but to make  `NoUnexpectedThunks` shut up about it any thunks in it, so that it can find the true space leaks.